modules/network/src/test/java/org/apache/ignite/network/DefaultMessagingServiceTest.java

```diff
@@ -208,10 +222,131 @@ public void sendMessagesTwoChannels() throws Exception {
+            );
+
+            operation.sendAction.send(senderServices.messagingService, TestMessageImpl.builder().build(), receiverNode);
+
+            assertTrue(
+                    waitForCondition(() -> channelId.get() != Integer.MAX_VALUE, TimeUnit.SECONDS.toMillis(10)),
+                    "Did not get any message in time"
+            );
+
+            assertThat(channelId.get(), is((int) operation.expectedChannelType.id()));
+        }
+    }
+
+    @ParameterizedTest
+    @EnumSource(RespondOperation.class)
+    void respondIsHandledInThreadCorrespondingToChannel(RespondOperation operation) throws Exception {
+        try (
+                Services senderServices = createMessagingService(senderNode, senderNetworkConfig);
+                Services receiverServices = createMessagingService(receiverNode, receiverNetworkConfig)
+        ) {
+            CountDownLatch invokeFuturePrepared = new CountDownLatch(1);
+
+            receiverServices.messagingService.addMessageHandler(
+                    TestMessageTypes.class,
+                    (message, sender, correlationId) -> {
+                        try {
+                            invokeFuturePrepared.await(10, TimeUnit.SECONDS);
+                        } catch (InterruptedException ignored) {
+                            // No-op.
+                        }
+
+                        operation.respondAction.respond(
+                                receiverServices.messagingService,
+                                message,
+                                Objects.requireNonNull(topologyService.getByConsistentId(sender)),
+                                Objects.requireNonNull(correlationId)
+                        );
+                    }
+            );
+
+            AtomicInteger responseChannelId = new AtomicInteger(Integer.MAX_VALUE);
+
+            CompletableFuture<NetworkMessage> invokeFuture = senderServices.messagingService.invoke(
+                    receiverNode,
+                    operation.expectedChannelType,
+                    TestMessageImpl.builder().build(),
+                    10_000
+            );
+            CompletableFuture<NetworkMessage> finalFuture = invokeFuture.whenComplete((res, ex) -> {
+                responseChannelId.set(extractChannelIdFromThreadName(Thread.currentThread()));
+            });
+
+            // Only now allow the invoke response to be sent to make sure that our completion stage executes in the channel's
+            // inbound thread.
+            invokeFuturePrepared.countDown();
+
+            assertThat(finalFuture, willCompleteSuccessfully());
+
+            assertThat(responseChannelId.get(), is((int) operation.expectedChannelType.id()));
+        }
+    }
+
+    private static int extractChannelIdFromThreadName(Thread thread) {
+        Pattern pattern = Pattern.compile("^.+-(\\d+)-\\d+");
```

Comment:
What's the second number after channel id? 0? Just asking

Reply:
Yes, the second number is the ordinal number of the thread in its pool, and it's 0 as we use single-thread executors.